### ISSUE TITLE
Call to Timber::get_widgets  kills wordpress

### ISSUE THREAD

**Issue by Jakkkub** [timber/timber#977]
### Expected behavior
This code should load data for three widget areas:

```
add_filter('timber_context', 'add_to_context');
function add_to_context($data){ 
    //footer widgets
    for($i = 1; $i<4; $i++){ 
        $data['footer_widgets_'.$i] = Timber::get_widgets('footer_widgets_'.$i);
    }
}
```

Code is places in timber_context filter

### Actual behavior
Nothing is loaded in browser.
After some time I see just Chrome`s ERR_CONNECTION_RESET page.
If I comment out line with Timber::get_widgets then it works.

It is fresh WP installation so there are no custom widgets - just WP`s default ones, so there should be no problem inside widgets.

I have used this snippet of code on many themes before, always worked. Maybe 1.0 upgrade issue?

### What version of WordPress, PHP and Timber are you using?
WordPress 4.5.1, PHP 7.0.1, Timber 1.0 -->

### How did you install Timber? 
From wp.org

**Comment by connorjburton** [comment]
Hi @Jakkkub - did this happen pre `1.0.0`?

**Comment by connorjburton** [comment]
Replicated this without the `add_filter` part and it works okay.

**Comment by elivz** [comment]
I'm seeing something similar. With `Timber::get_widgets` inside the `timber_context` action callback, it appears to be creating an infinite recursion. I see the first widget area repeated many times on the page (and no other content), and then PHP gives up with `Fatal error: Maximum function nesting level of '256' reached, aborting!`. Let me know if there is anything I can do to help track this down - I can post the stack trace if that would be helpful.

**Comment by connorjburton** [comment]
Hi, yes the stack trace would be useful. Thanks

**Comment by connorjburton** [comment]
Hi guys,

I am doing the following and i'm not getting any errors, a full stack trace would be useful along with your guys full setup in terms of what is in your widgets.

functions.php
```php
register_sidebars(2, array('name'=>'foobar%d'));

	add_filter('timber_context', 'add_to_context');
	function add_to_context($data){
		for($i = 1; $i<2;$i++){ 
	        $data['foobar'.$i] = Timber::get_widgets('foobar'.$i);
	    }

		return $data;
	}
```

**Comment by elivz** [comment]
Ah, I think I figured out what is going on. One of my sidebar widgets pulls in a template file from my theme, which called `Timber::get_context()`, which then loads the widgets again, and round and round. I'm not sure what changed between 0.22.5 and 1.0.1, but it looks like Timber must have identified that situation somehow before and prevented the loop. (I didn't change anything in my templates or widgets between when it worked and when it stopped - I just upgraded Timber).

**Comment by Jakkkub** [comment]
Hi, I did more investigation today: I removed all widgets (worked) and tried to add them back one by one. Problematic one was Search widget - and it uses this custom code in searchform.php with `Timber::get_context` call:

```
<?php
$data = Timber::get_context();

if(isset($_GET['s'])){
	$data['searchterm'] = $_GET['s'];
}

Timber::render('searchform.twig', $data);
```
When I used the same theme with pre-1.0 Timber, it worked.

So it seems that elivz is right.

**Comment by connorjburton** [comment]
Hmm, yeah calling get context in get context obviously won't work, but it's interesting `1.0.1` exposed it. I'll have a look guys, but I would recommend fixing the infinite loops, obviously.

This might not be something we will push a fix for necessarily.

**Comment by Jakkkub** [comment]
So `Timber::get_context()` "recreates" the context on every call?
 
So what is the best way in such cases to access context data (like in searchform.php, which may be included by Search widget as well as in page templates etc.)?

**Comment by connorjburton** [comment]
Well it's not that `Timber::get_context()` isn't the best way to do what you want, it's just that it's going to cause an infinite loop, because you are calling `get_context` that then does the action, which then calls get widgets, which calls `get_context`, ad infinitum.

I will get back to you on this as need to think about best way

**Comment by elivz** [comment]
It looks like the solution is to replace `$context = Timber::get_context();` with `$context = $GLOBALS['timberContext'];` in any templates that will only ever be included from other templates (as opposed to called on their own). I noticed that global was being set in the starter theme's header.php template. I am guessing - though I haven't tested - that this will give a bit of a performance boost too, since I'm not regenerating all the sidebars and menus several times per request.

For reference, here is my new `searchform.php`:

    $context = $GLOBALS['timberContext'];
    $context['id'] = uniqid();
    Timber::render( 'search-form.twig', $context );

**Comment by Jakkkub** [comment]
> $context = $GLOBALS['timberContext'];
Nice idea... Maybe there could be some nice way built-in, like Timber::context property or something?
Or make Timber::get_context() working something like that?
```
public function get_context(){
  if ($context does not exist yet) {
    create and return $context
  }else{
    return existing $context
  }
}
```
(which is the way I supposed it works)

**Comment by connorjburton** [comment]
Even if we put that check in `get_context` it would still cause an infinite loop as you are still calling the function.

I am puzzled how it worked before, hadn't had chance to look at this yet but will be fixed/solution provided in `1.0.2`

**Comment by connorjburton** [comment]
Hey guys, please see attached PR. Can you guys give this a go and see if it solves the original issue (without any code changes to fix the issue).

Either just copy the code (easiest) or require the branch with composer

**Comment by aaronhipple** [comment]
We're experiencing issues with this as well.

@connorjburton Your PR seems to address the loop however we're still getting odd behavior with calls to get_widgets (some of which are inside our timber_context filter functions). At least some widgets seem to be rendering too early (perhaps on loading the context?) causing something like this at the top of the output:

```html
<div id="fsm_content_block_widget-8" class="widget_fsm_content_block_widget">
    <!-- Widget Content -->
</div>
<div id="fsm_brand_widget_block-2" class="widget_fsm_brand_widget_block">
    <!-- More Widget Content -->
<!doctype html>
    <!-- template content follows -->
```

May be unrelated to the looping issue. I'll dig for more info as I'm able.

**Comment by aaronhipple** [comment]
@connorjburton Playing with our issue a bit more, it does seem that the loop is fully resolved by your pull request and that our remaining issues are coming from elsewhere (Seems like output buffering in TimberHelper::function_wrapper() is behaving oddly in some circumstances -- I'll investigate further and address that separately)

**Comment by Jakkkub** [comment]
Your PR solved the issue for me, everything works so far

**Comment by connorjburton** [comment]
Okay cool, there are a couple of problems with that PR so don't take it as gospel, just a proof of concept on how we can fix it. Follow that PR for updates.

**Comment by connorjburton** [comment]
Merged and will be released in `1.0.2`

**Comment by elivz** [comment]
Sorry, just finally got a chance to test out your fix. It works perfectly in my setup now too. Thanks!
